src/tech/v3/dataset/impl/column_index_structure.clj

```diff
@@ -0,0 +1,86 @@
+(ns tech.v3.dataset.impl.column-index-structure
+  (:import [java.util TreeMap LinkedHashMap]
+           [tech.v3.datatype ListPersistentVector])
+  (:require [tech.v3.protocols.column :as col-proto]
+            [tech.v3.datatype :refer [elemwise-datatype clone ->buffer]]
+            [tech.v3.datatype.argops :refer [arggroup]]
+            [tech.v3.datatype.casting :refer [datatype->object-class]]
+            [clojure.set :refer [difference]]))
+
+
+(extend-type TreeMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      ::pick
+      (let [^TreeMap new-index-structure (.clone ^TreeMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure)
+      ::slice
+      (let [{from            :from
+             from-inclusive? :from-inclusive?
+             to              :to
+             to-inclusive?   :to-inclusive?} selection-spec]
+        (.subMap ^TreeMap index-structure
+                 from
+                 (if (nil? from-inclusive?) true from-inclusive?)
+                 to
+                 (if (nil? to-inclusive?) true to-inclusive?))))))
+
+
+(extend-type LinkedHashMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      ::pick
+      (let [^LinkedHashMap new-index-structure (.clone ^LinkedHashMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure))))
+
+
+;; https://github.com/techascent/tech.ml.dataset/blob/master/src/tech/v3/dataset/impl/column.clj#L41
+(defn categorical? [dtype]
+  (boolean (#{:string :keyword :symbol} dtype)))
```

Comment:
I'm a bit concerned this is brittle because what's considered categorical could change elsewhere (e.g.  [here](https://github.com/techascent/tech.ml.dataset/blob/master/src/tech/v3/dataset/impl/column.clj#L41)) and someone could forget to change it in this ns. Could this be made more central? A function in dtype-next?

Reply:
I think another definition that is likely to be valid is any datatype that is not a boolean or a number is categorical.  It has to be converted to a number in order to use machine learning on it so the result is the same in a lot of senses.  But yes, I think it is fragile in that if you 'assoc' a column that has categorical information on it then that flag isn't set.  I kind of think that information shouldn't be set by `tmd` at all but set by metamorph when needed.  @behrica - What do you think about `tmd` not setting the `categorical?` metadata information on the column?

Reply:
Interesting. Just to make sure I understood what you said in the first sentence, the essence of it is this part: "any datatype that is not a boolean or a number is categorical"? Is that true, though? What about time data? That could be categorical in some circumstances, but in others as in when you have a time index certainly  not, right? 

Re: the implementation here for generating the index, what I tried to do is allow the `:categorical?` metadata property to override, any attempt to determine the type based on the datatype using this `categorical?` function. The logic is in the dispatch function for the `make-index-structure` multi method. ([here](https://github.com/techascent/tech.ml.dataset/pull/214/files#diff-85f732aa1acc3681a0091e4e7da3fa81e019ff6d0212c922e0093e495157407aR60-R71)). Basically, if the `:categorical?` metadata is present, and it's `true` the method dispatches the `::categorical` keyword; otherwise, it dispatches the object class of the datatype. Alternately, if the `:categorical?` metadata is not present, it checks to see if the data's datatype can be considered categorical using the `categorical?` method here (that mimicks the logic used to set the categorical flag in `new-column`); otherwise, it dispatches the object class of the datatype. When the multimethod dispatch function is the object class, that will be matched by the default multimethod defn, unless the user is working with a library that has "extended" TMDs index by defining a new multimethod for some particular class or set of classes (e.g. geometries). 

I guess, even if TMD doesn't set the categorical metadata we'd still want to try to determine if the data's datatype can be considered categorical? Or @cnuernber are you suggesting that we don't ever try to determine if the datatype can be considered categorical, and only dispatch a categorical index structure if the user (or a higher level library like metamorph) has explicitly set a `:categorical?` metadata? 

If that's what you mean, it would certainly be simplifying in this context and I can see it in the sense that then TMD takes a kind of stance of neutrality as it were, which is at least predictable. One thing I'm wondering is how this question fits conceptually with the more opinionated stance that TMD will now take toward setting the datatype flag on the metadata in the `6.00-beta-1`? I guess the point is that it's hard to define in any clear way what is categorical or not, and it may depend on the intent of the user? (Henrique Nunes posted this stack exchange post discussion related to this topic, but I still go the sense here that there's no solid way to dynamically make the determination: [How can I dynamically distinguish between categorical data and numerical data?](https://datascience.stackexchange.com/questions/9892/how-can-i-dynamically-distinguish-between-categorical-data-and-numerical-data))

Since I'm hoping to wrap this PR soon, do we think what's here is reasonable as is, or do we maybe want to make a few changes. I could definitely see here removing the attempt to determine the type dynamically and only leaving the check on the `:categorical?` flag. Then if you ulimately also decide to remove that elsewhere and a library like `metamorph` starts adding the flag we can still support that so long as we know what the flag will be...

Reply:
What I am thinking is that the categorical? flag on columns is error prone because if something like column-map happens the metadata may not reflect that this is now a categorical column or something like that.

time-based datatypes have a non-user-defined conversion to numbers but out of the range of possible user-specified datatypes which is infinite this is a super special case.  Columns containing tensors or persistent maps or vectors may not be `categorical?` in a normal sense but they do have some of the properties namely the best automatic map of object->number is a categorical-type map with hash keys being object instances and values being incrementing integers.

That is really all I am getting at is that whether something is categorical or not is such an odd distinction that really all we can say is if there is a defined conversion to a scalar number or not.  In the general case of possible objects the only sane automatic conversion is a dictionary-based conversion of some sort.  And as noted before the distinction goes the other way as well as numeric columns sometimes stand for categorical values and should be treated as such if they are the target of inference or if they are used in statistical operations (mode instead of mean, for instance).

So assigning categorical to metadata like I did I think isn't the best way forward as the general case of conversion to a numeric value cannot be done repeatably across the infinite range of user-defined datatypes without user intervention of some sort.

Reply:
This is helpful. Thanks!

Reply:
I see that from #236 there is a new function `column-datatype-categorical?` that is used in `new-column` to decide whether to set the `:categorical?` flag (unless it's been set already). I can't import that here b/c it creates a cyclical dependency (though I suppose it could be placed elsewhere), but I'm wondering if given the auditing of the type checking from that PR, it's safe to just check the `categorical?` meta below in the dispatch fn ([here](https://github.com/techascent/tech.ml.dataset/pull/214/files#diff-85f732aa1acc3681a0091e4e7da3fa81e019ff6d0212c922e0093e495157407aR66-R72)) b/c we can rely on that flag having been set where it needs to have been...

Reply:
Honestly I would rather move the function and given our earlier conversation that function is incorrect; it will mark datetime types as categorical.  

I think for now you can assume that categorical? will be set as accurately as it was before.  We can look at removing it or addressing things more thoroughly at a later time.  I will move that function somewhere that will not create cyclic dependencies.

Reply:
Moved `column-datatype-categorical?` to [column-base.clj](https://github.com/techascent/tech.ml.dataset/blob/eaa79171ade1039d12f8a860cbdae38f97cabe8c/src/tech/v3/dataset/impl/column_base.clj#L84).

Reply:
Wonderful. Thank you.

Reply:
I just cleaned up the definition to something that has meaning.  I am sure we will revisit the categorical/not categorical topic.  The next step I think is tracking statistics during load or any time we have to scan data so we know min,max and possibly for numeric types the unique set up to a set low threshold (like 32).  Then if we are loading strings where we also keep track of the unique set already we autodetect when that isn't a great idea, like when we have read 1000 rows and have >900 unique entries.

That (per-column statistics/heuristics upon load) is for another PR :-).